### PR TITLE
OMN Token Bridge

### LINKED ISSUE #1815: Deposit/Withdraw for ERC-20 token(OMN) - Design Specification
Deposit/Withdraw:
deposit:withdraw-allowance

small restyled connection modal:
Connection

Notes: 
- We should use the same technology used on https://omni.xdaichain.com/bridge
- Design specification have paddings/sizes for the modals changed, please take a closer look and style accordingly
- Deposit/Withdraw modal has now a list of available assets which can be picked via radio buttons

Please provide feedback/ask questions if something is unclear.

**Figma**:
[Design Specification](https://www.figma.com/file/aTivxuHmp2l6Sn8nho1Ngu/Playground?node-id=2978%3A2475)
[Playground](https://www.figma.com/proto/aTivxuHmp2l6Sn8nho1Ngu/Playground?node-id=2978%3A2475&viewport=-1027%2C272%2C0.17787669599056244&scaling=min-zoom&page-id=2978%3A2386)

### PR DESCRIPTION
**Pull request by Mi-Lan**
How to Test:
- Since Omen Token is not deployed i used GEN erc-20 in the background...Get some gen
Deposit:
- Go into Your connection and select deposit
- Then approve Omen Token bridge
- Desposit some amount 
Withdraw:
- Go into your connection modal
- Select withdraw
- Enter some amount 
- Since claim option is not scope of this pr only way to confirm if transaction is successful is to go into omni bridge ui and confirm heres link https://omni.xdaichain.com/history

### CODE REVIEW

**Review by pimato — CHANGES_REQUESTED** [review]
@Mi-Lan I am getting a compile warning:
```Compiled with warnings.

./src/hooks/useCollateralBalance.tsx
  Line 35:5:  Unexpected console statement  no-console```

**Comment by pimato** [comment]
Margin between the Inputfield and the required amount info message should be 16px and not 12px:
Bildschirmfoto 2021-04-21 um 09 59 44

**Comment by pimato** [comment]
After closing the confirmed transaction modal I went to the `Your Connection - Modal` and it seems like the balance is not getting updated after we did a successful deposit:
Bildschirmfoto 2021-04-21 um 11 11 22

**Comment by pimato** [comment]
Once the approval has been set we should not remove it but rather show it like we do on Market creation process:
Bildschirmfoto 2021-04-21 um 11 23 29

**Comment by pimato** [comment]
looks like I can`t withdraw my OMN tokens with the max button: 
Bildschirmfoto 2021-04-21 um 11 31 18

**Comment by pimato** [comment]
Instead of showing the transaction confirmed modal and user needs to get back to the `Your Connection - Modal` manually, we should transition from `Transaction modal - Pending Confirmation State` to the `Your Connection modal` directly with an updated OMN balance so the user sees the direct result of his action.

**Comment by Mi-Lan** [comment]
> After closing the confirmed transaction modal I went to the `Your Connection - Modal` and it seems like the balance is not getting updated after we did a successful deposit:
> Bildschirmfoto 2021-04-21 um 11 11 22
> 
> after refresh:
> Bildschirmfoto 2021-04-21 um 11 12 30

Hmm yeah seems that's the case for Dai also i will see to optimize deposit/withdraw function to encompass it...

**Comment by hexyls** [comment]
Hey man, I deployed a new [MultiClaim contract](https://etherscan.io/address/0xbbc9f0b5f1243946c42cdd6bc2267604b4f9a313#code) for you and you can just pass in an additional bridge address like you suggested.

I had a poke around the omnibridge docs just so I could get a better understanding. Those docs are so outdated lol anyway I eventually discovered that there is a [third bridge address](https://etherscan.io/address/0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e#readProxyContract) for the return trip from xdai->mainnet and this is where we send claim transactions to.

So bridge addresses are:

eth -> xdai bridge: 0x88ad09518695c6c3712AC10a214bE5109a655671
xdai -> eth bridge: 0xf6A78083ca3e2a662D6dd1703c939c8aCE2e268d
claim to wallet: 0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e

**Comment by Mi-Lan** [comment]
> MultiClaim contract

@hexyls Awesome man! Thank you! But I will include multi claim in separate pr...

**Review by pimato — CHANGES_REQUESTED** [review]
Bildschirmfoto 2021-04-22 um 14 37 27
Bildschirmfoto 2021-04-22 um 14 37 22

Asset Description should be capitalized.

**Comment by pimato** [comment]
I want to enable Dai on a new Account and it seems like I can`t load the asset icon: 
Bildschirmfoto 2021-04-22 um 14 51 53

Even though the link works just fine:
https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x6B175474E89094C44Da98b954EedeAC495271d0F/logo.png

**Comment by pimato** [comment]
Remove decimals for any asset in the info box:
Bildschirmfoto 2021-04-28 um 16 49 09

**Comment by pimato** [comment]
Reset Inputfield on Asset switch:
Bildschirmaufnahme 2021-04-28 um 16 57 11

Currently it anything the user has put in.

**Comment by pimato** [comment]
We need to explain the user why the withdrawal of ERC-20 tokens does not work if the users omen account is not funded with Dai:
Withdraw

Note: 
- Replace min withdrawal amount message with the message above.

Figma:
[Design Specification](https://www.figma.com/file/R056tSYMjvfwaFidEBItYy/Widgets-V2?node-id=3832%3A241)

**Review by pimato — CHANGES_REQUESTED** [review]
After I deposited/withdrew funds the appropriate balances have not been updated in the my connection modal.

**Comment by Mi-Lan** [comment]
@KadenZipfel @hexyls There is some problem with the fetching of balance for xOmen even after refreshing couple of times it doesn't update....Would be good if you guys can provide some feedback. I haven't been able to figure it out.
in useConnectedBalance I also tried to get balance with useCollateral but it was doing infinite loop for some reason i even left comments there...

**Comment by hexyls** [comment]
Hey @Mi-Lan, I took a look at this - the infinite loop in `useCollateralBalance` was caused by `collateral` in the dependency array. The `collateral` value comes from `getToken` which returns a new object with each render. I updated the dependency array in `useCollateralBalance` to use `collateral.symbol` instead of the whole object which sorted it out.

The other issue was token balances not updating, this turns out to be because the validators are really, really slow on the Omni bridge. Waiting 9 confirmations was not guaranteeing that tokens had been relayed, so when we refetched balances they hadn't actually changed yet. I added `waitForBridgeMessageStatus` that will wait until the Omni bridge validators have actually processed the deposit, so when we fetch balances after a tx we can be sure they are up to date.

**Comment by Mi-Lan** [comment]
> Hey @Mi-Lan, I took a look at this - the infinite loop in `useCollateralBalance` was caused by `collateral` in the dependency array. The `collateral` value comes from `getToken` which returns a new object with each render. I updated the dependency array in `useCollateralBalance` to use `collateral.symbol` instead of the whole object which sorted it out.
> 
> The other issue was token balances not updating, this turns out to be because the validators are really, really slow on the Omni bridge. Waiting 9 confirmations was not guaranteeing that tokens had been relayed, so when we refetched balances they hadn't actually changed yet. I added `waitForBridgeMessageStatus` that will wait until the Omni bridge validators have actually processed the deposit, so when we fetch balances after a tx we can be sure they are up to date.

Thank you @hexyls 🤗

File: app/src/components/common/icons/IconOmen.tsx [hunk 1]
```diff
@@ -1,50 +1,44 @@
...
 
-export const IconOmen = () => {
+interface Props {
+  style?: any
+  size?: number
+  dropShadow?: boolean
+}
+
+const SvgStyling = styled.svg<{ dropShadow?: boolean }>`
+  ${props => props.dropShadow && 'filter: drop-shadow(3px 3px 2px rgba(0, 0, 0, 0.7));'};
```
Comment (pimato) [inline app/src/components/common/icons/IconOmen.tsx:11]: lets use: filter: drop-shadow(0px 2px 4px rgba(13, 71, 161, 0.25)

**Review by hexyls — APPROVED** [review]
🙌 looks good!

File: app/src/services/xdai.ts [hunk 2]
```diff
@@ -135,15 +192,42 @@ class XdaiService {
...
+    let status
+    while (!status) {
+      // get tx receipt from mainnet
+      const transaction = await provider.signer.signer.provider.getTransactionReceipt(hash)
+      if (transaction) {
+        // get messageId from transaction logs
+        const log = transaction.logs[transaction.logs.length - 1]
+        const messageId = log.topics[log.topics.length - 1]
+        // check if the message has been processed
+        const xdaiOmniAddress = '0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59'
```
Comment (KadenZipfel) [inline app/src/services/xdai.ts:224]: This address should be a constant right?

**Review by kadenzipfel — CHANGES_REQUESTED** [review]
Looking really good! Just some small suggestions here. I'm gonna do some proper testing in my next run through when I hopefully have more time

File: app/src/components/modal/modal_deposit_withdraw/index.tsx [hunk 3]
```diff
@@ -84,45 +121,119 @@ export const ModalDepositWithdraw = (props: Props) => {
...
+  const minDaiExchange = exchangeType === ExchangeType.deposit ? parseEther('5') : parseEther('10')
+  const minOmenExchange = exchangeType === ExchangeType.deposit ? parseEther('1') : parseEther('1')
   const isDepositWithdrawDisabled =
-    displayFundAmount.isZero() || !wallet || displayFundAmount.gt(wallet) || displayFundAmount.lt(minExchangeAmount)
+    displayFundAmount.isZero() ||
+    !wallet ||
+    displayFundAmount.gt(wallet) ||
+    displayFundAmount.isZero() ||
+    !wallet ||
+    displayFundAmount.gt(wallet) ||
```
[the remark this answers is not in the document]
Comment (Mi-Lan) [reply app/src/components/modal/modal_deposit_withdraw/index.tsx:195]: good catch!

File: app/src/services/xdai.ts (hunk 2 above)
Comment (Mi-Lan) [reply app/src/services/xdai.ts:224]: Dunno how that slipped! Good catch

**Review by pimato — APPROVED** [review]
great work!

**Review by kadenzipfel — APPROVED** [review]
Haven't tested, but code looks 👌

**Review by Scheusal13 — COMMENTED** [review]
0xD4EBBC1fE630e1aF1e2143d11b2C19c48d099098

**Pull request merged**